### PR TITLE
Movies and people module and API

### PR DESCRIPTION
**Pull request by gabrieelgomez**
Add unit test with Rspec
Add Movies and People models - relationships

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,7 +1,9 @@
+# frozen_string_literal: true
+
 source 'https://rubygems.org'
 git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 
-ruby '2.7.1'
+ruby '2.6.5'
```
Comment (enmachs) [inline Gemfile:6]: why this specific version?

File: app/controllers/api/v1/api_controller.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+# Api::V1 module
+module Api
+  # V1 module
+  module V1
+    # ApiController controller
+    class ApiController < ApplicationController
+      respond_to :json
+
```
Comment (enmachs) [inline app/controllers/api/v1/api_controller.rb:10]: needed empty line?

File: app/controllers/api/v1/movies/movies_controller.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
```
Comment (enmachs) [inline app/controllers/api/v1/movies/movies_controller.rb:1]: Due to the scope of changes required, i suggest to move `movies_controller.rb` out of `movies` folder. Since I don't see necessary of that Movie module

File: app/controllers/api/v1/movies/movies_controller.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    private
+
+    def movie_params
+      params.require(:movie).permit(
+        :title, :release_year, person_ids: []
+      )
+    end
+
+    def set_movie
+      @movie = Movie.find_by(id: params[:movie_id])
```
Comment (enmachs) [inline app/controllers/api/v1/movies/movies_controller.rb:53]: what happen in case record is not found? (destroy and update will fail)

File: app/controllers/api/v1/movies/movies_controller.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+# Api::V1 module
+module Api::V1
+  # Movies::MoviesController controller
+  class Movies::MoviesController < ApiController
+    before_action :authenticate_v1_user!, except: %i[index show]
+    before_action :set_movie, only: %i[show updatedestroy]
```
Comment (enmachs) [inline app/controllers/api/v1/movies/movies_controller.rb:8]:
```suggestion
    before_action :set_movie, only: %i[show update destroy]
```

File: app/controllers/api/v1/people/people_controller.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+
+# Api::V1 module
+module Api::V1
+  # People::PeopleController controller
+  class People::PeopleController < ApiController
+    before_action :authenticate_v1_user!, except: %i[index show]
```
Comment (enmachs) [inline app/controllers/api/v1/people/people_controller.rb:7]: Requirements demand authenticated users for "unsafe methods". Here need to add them to `:authenticate_v1_user!` before_action. (same for MoviesController)

File: app/models/concerns/roleable.rb [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+  extend ActiveSupport::Concern
+
+  included do
+    def has_role?(role_name)
+      role_array = roles.where(name: role_name)
+
+      return false if role_array.empty?
+
+      role_array != []
+    end
```
Comment (enmachs) [inline app/models/concerns/roleable.rb:14]:
```suggestion
    def has_role?(role_name)
      role_array = roles.where(name: role_name)

      role_array.any?
    end
```

File: app/models/movie.rb [hunk 8]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+# Movie class
+class Movie < ApplicationRecord
+  has_and_belongs_to_many :people
+
+  validates :title, :release_year, presence: true
+
+  def people_as(role_name)
```
Comment (enmachs) [inline app/models/movie.rb:9]: please add comments. References to what does this method and other similars

File: app/models/person.rb [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+  include Roleable
+
+  has_and_belongs_to_many :roles
+  has_and_belongs_to_many :movies
+
+  accepts_nested_attributes_for :people_roles
+
+  validates :first_name, :last_name, :aliases, presence: true
+  validates :genre,
+            inclusion: %w[male female],
```
Comment (enmachs) [inline app/models/person.rb:14]: You can use a constant to this rule

File: app/models/person.rb [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+            allow_nil: false
+
+  def movies_as(role_name)
```
Comment (enmachs) [inline app/models/person.rb:17]: I suggest to validate that person actually owns that role, using `has_role`, before making db query search. (this would save db time)

File: app/models/user.rb [hunk 11]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+# User class
+class User < ActiveRecord::Base
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, :registerable,
```
Comment (enmachs) [inline app/models/user.rb:7]: do you need all these devise modules? if not, remove the ones you're going to use

File: app/controllers/api/v1/movies/movies_controller.rb (hunk 3 above)
Comment (gabrieelgomez) [reply app/controllers/api/v1/movies/movies_controller.rb:1]: Ready

File: app/controllers/api/v1/api_controller.rb (hunk 2 above)
Comment (gabrieelgomez) [reply app/controllers/api/v1/api_controller.rb:10]: Ready

File: app/controllers/api/v1/movies/movies_controller.rb (hunk 5 above)
Comment (gabrieelgomez) [reply app/controllers/api/v1/movies/movies_controller.rb:8]: Ready

File: app/models/role.rb [hunk 12]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+# Role class
+class Role < ApplicationRecord
```
Comment (enmachs) [inline app/models/role.rb:4]: where the roles are created? I assume that when a new movie is created and persons are to it as "janitor man" of instance. Does this logic exist?

File: app/models/concerns/roleable.rb (hunk 7 above)
Comment (gabrieelgomez) [reply app/models/concerns/roleable.rb:14]: ready

File: config/database.yml [hunk 13]
```diff
@@ -80,6 +85,6 @@ test:
 #
 production:
```
Comment (enmachs) [inline config/database.yml:86]: suggest to use secret or env vars when deploy

File: app/models/movie.rb (hunk 8 above)
Comment (gabrieelgomez) [reply app/models/movie.rb:9]: ready

File: app/models/person.rb (hunk 9 above)
Comment (gabrieelgomez) [reply app/models/person.rb:14]: ready, GENRES = %w[male female].freeze

File: config/initializers/routing_draw.rb [hunk 14]
```diff
@@ -0,0 +1,5 @@
+class ActionDispatch::Routing::Mapper
+  def draw(routes_name)
```
Comment (enmachs) [inline config/initializers/routing_draw.rb:2]: good one (but maybe amounts of routes is not the big deal)

File: config/routes/people.rb [hunk 15]
```diff
@@ -0,0 +1,7 @@
+namespace :people do
```
Comment (enmachs) [inline config/routes/people.rb:1]: maybe an endpoint to the movies from an specific people as actor?

File: db/migrate/20200425145743_create_people.rb [hunk 16]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+# CreatePeople migration
+class CreatePeople < ActiveRecord::Migration[6.0]
+  def change
+    create_table :people do |t|
+      t.string :first_name
```
Comment (enmachs) [inline db/migrate/20200425145743_create_people.rb:7]: default values? null allowed?

File: db/migrate/20200425164406_create_movies.rb [hunk 17]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+# CreateMovies migration
+class CreateMovies < ActiveRecord::Migration[6.0]
+  def change
+    create_table :movies do |t|
```
Comment (enmachs) [inline db/migrate/20200425164406_create_movies.rb:6]: same comment for defaults and null

File: db/seeds.rb [hunk 18]
```diff
@@ -5,3 +5,33 @@
 #
 #   movies = Movie.create([{ name: 'Star Wars' }, { name: 'Lord of the Rings' }])
 #   Character.create(name: 'Luke', movie: movies.first)
+
+User.create(
+  email: 'user@admin.com',
```
Comment (enmachs) [inline db/seeds.rb:10]: puts will add more status output of this process

File: spec/factories/user.rb [hunk 19]
```diff
@@ -0,0 +1,7 @@
+FactoryBot.define do
+  factory :user do
+    email { "admin@mail.com" }
+    password { "12345678"}
+    password_confirmation { "12345678"}
```
Comment (enmachs) [inline spec/factories/user.rb:5]:
```suggestion
    password { "12345678" }
    password_confirmation { "12345678" }
```

File: test/fixtures/movie_people.yml [hunk 20]
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (enmachs) [inline test/fixtures/movie_people.yml:1]: Fixtures were generated by default, are they going to be used ? I see you got faker for generation task. If they're not required you can remove them

File: spec/controllers/api/v1/movies_controller_spec.rb [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+        before do
+          create_list(:movie, 1)
+          get :show
+        end
+
+        it 'returns a ok status' do
+          expect(response).to have_http_status(:ok)
+        end
+
+        it 'responds with a non empty list' do
```
Comment (enmachs) [inline spec/controllers/api/v1/movies_controller_spec.rb:48]: test when db is empty

File: spec/controllers/api/v1/movies_controller_spec.rb [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+
+        before do
+          post :create, params: movie_params
+        end
+
+        it 'responds with status created' do
+          expect(response).to have_http_status(:created)
+        end
+
+        it 'responds with the object created' do
```
Comment (enmachs) [inline spec/controllers/api/v1/movies_controller_spec.rb:31]: add wrong cases to tests, what is the expected when a movie is not found or when a movie is not created just fine

File: spec/controllers/api/v1/movies_controller_spec.rb [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+        before do
+          create_list(:movie, 10)
+          get :index
+        end
+
+        it 'returns a ok status' do
+          expect(response).to have_http_status(:ok)
+        end
+
+        it 'responds with a non empty list' do
```
Comment (enmachs) [inline spec/controllers/api/v1/movies_controller_spec.rb:65]: also test when is empty

File: spec/controllers/api/v1/people_controller_spec.rb [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+        let(:person) { create(:person) }
+
+        let(:person_params) do
+          {
+            'person': {
+              'first_name': Faker::DcComics.name,
+              'last_name': Faker::Name.last_name,
+              'aliases': Faker::Internet.username,
+              'role_ids': Role.shuffle[1..5].map(&:id)
+            }
```
Comment (enmachs) [inline spec/controllers/api/v1/people_controller_spec.rb:84]: this can come from fixtures

File: spec/controllers/api/v1/people_controller_spec.rb [hunk 25]
```diff
@@ -0,0 +1,99 @@
+require 'rails_helper'
+
+module Api::V1
+  # Peole::PeoleController controller
+  RSpec.describe Peole::PeoleController, type: :controller do
```
Comment (enmachs) [inline spec/controllers/api/v1/people_controller_spec.rb:5]: Same comment of movies, cover also wrong scenarios
